// @substrate/api-sidecar/build/src/services/transaction/TransactionDryRunService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionDryRunService = void 0;
const responses_1 = require("../../types/responses");
const AbstractService_1 = require("../AbstractService");
const extractCauseAndStack_1 = require("./extractCauseAndStack");
class TransactionDryRunService extends AbstractService_1.AbstractService {
    async dryRuntExtrinsic(hash, transaction) {
        const { api } = this;
        try {
            const [applyExtrinsicResult, { number }] = await Promise.all([
                api.rpc.system.dryRun(transaction, hash),
                api.rpc.chain.getHeader(hash),
            ]);
            let dryRunResult;
            if (applyExtrinsicResult.isOk) {
                dryRunResult = {
                    resultType: responses_1.TransactionResultType.DispatchOutcome,
                    result: applyExtrinsicResult.asOk,
                };
            }
            else {
                const { asError } = applyExtrinsicResult;
                dryRunResult = {
                    resultType: responses_1.TransactionResultType.TransactionValidityError,
                    result: asError.isInvalid
                        ? asError.asInvalid
                        : asError.asUnknown,
                    validityErrorType: asError.isInvalid
                        ? responses_1.ValidityErrorType.Invalid
                        : responses_1.ValidityErrorType.Unknown,
                };
            }
            return {
                at: {
                    hash,
                    height: number.unwrap().toString(10),
                },
                dryRunResult,
            };
        }
        catch (err) {
            const { cause, stack } = extractCauseAndStack_1.extractCauseAndStack(err);
            throw {
                at: {
                    hash,
                },
                error: 'Unable to dry-run transaction',
                transaction,
                cause,
                stack,
            };
        }
    }
}
exports.TransactionDryRunService = TransactionDryRunService;
